File: Algorithms/Sorting/Counting_Sort.cpp
//Name : Shivam Singhal
//Email ID : shivamsinghal0610@gmail.com

#include <bits/stdc++.h>
using namespace std;
 
void countSort(long arr[],long n){
    long l = 0;
    for(long i=0; i<n ;i++){
        //finding largest element to hash
        if(arr[i]>l)
            l = arr[i]; 
    }
    //initialising count of each element present in original array to 0
    long count[l+1] = {0}; 
    //storing frequency of each element using hashing
    for(long i = 0; i<n ; i++)
        count[arr[i]]++; 
    //counter to keep track of position in original array
    long c = 0; 
    for(long i=0; i<=l ;i++){
        //signifies that element i is present in original array
        if(count[i]>0){ 
            while(count[i]!=0){ 
                //overwriting onto original array as many times as element is present
                arr[c++] = i; 
                count[i]--;
            }
        }
    }
}
 
void printArray(long arr[],long n){
    for(long i=0; i<n ;i++)
        printf("%lld ",arr[i]);
    printf("\n");
}

int main(){
    long n;
    printf("Please enter the number of elements."); 
    scanf("%lld",&n);
    printf("\nPlease enter elements to be sorted.\n");
    long arr[n];
    //make sure that the maximum element is lesser than the maximum number of elements an array can hold
    for(long i=0; i<n ;i++)
        scanf("%lld",&arr[i]); 
    printf("Entered elements:\n");
    printArray(arr,n);
    //function to sort elements in ascending order using counting sort technique 
    countSort(arr,n); 
    printf("Sorted array:\n");
    printArray(arr,n);
    return 0;
}

//Time Complexity : O(max). Its more accurately the max of array size and the max element in the array.
//Auxiliary Space Complexity : O(max)

File: Algorithms/Sorting/Java/CountingSort.java
//  Counting Sort in java

class CountingSort { 
  static  void sort(char arr[]) 
    { 
        int n = arr.length; 
  
       //Array to store the final sorted array.
        char out[] = new char[n]; 
  
       //A new array to store the number of occurrences of a character.
        int c[] = new int[256]; 
        for (int i = 0; i < 256; ++i) 
            c[i] = 0; 
  
        //Adding the count of each element.
        for (int i = 0; i < n; ++i) 
            ++c[arr[i]]; 
  
       //Stores real position of element.
        for (int i = 1; i <= 255; ++i) 
            c[i] += c[i - 1]; 
  
       //Building the final sorted array.
        for (int i = 0; i < n; ++i) { 
            out[c[arr[i]] - 1] = arr[i]; 

            //Decreasing the count of the element which is positioned correctly.
            --c[arr[i]]; 
        } 
  
       
        for (int i = 0; i < n; ++i) 
           arr[i] = out[i]; 
    } 
  
    // Driver method 
    public static void main(String args[]) 
    { 
       
        char a[] = {'a','f', 'd','j','w','i'}; 
  
        //sort function is called which sorts the array.
        sort(a);
                    
  
        //Printing the sorted array.
        for (int i = 0; i < a.length; i++) 
            System.out.println(a[i]); 
    } 
} 

//Time Complexity -  O(n+k) where n is the number of elements in input array and k is the range of input.

//Space Comlexity: O(n+k)